Rubric support

ToeBee opened the pull request:
Rubrics are complicated and not everything is implemented but this at least allows:
* Reading of rubric objects from Canvas 
* Inclusion of rubric data when requesting assignment and submission objects

alina-spectorov commented:
Hi Toby, 
I am glad that you are working on this project again. I think your package is very good! 
I hope to review your rubrics pull request this week. Please let me know if it cannot wait.
Thank you again, Alina.

alina-spectorov commented:
1.	Thank you for fixing tabs and spaces.

2.	Your Rubrics objects look good to me. I hope I can still use the AssignmenReader and the SubmissionReader (with the RUBRIC_ASSESSMENT included in GetSubmissionsOptions) and will only need to make the following changes:
a.	Assignment.Rubric ->  RubricCriterion
b.	Assignment.Rating -> RubricRating
c.	Assignment.RubricSettings -> RubricSettings
d.	Submission.RubricAssessment -> Submission . SubmissionRubricAssessment

3.	For consistency, you may want to change Assignment.setRubric method name to Assignment.setRubricCriteria.

4.	RubricCriterion.java is missing the ignoreForScoring String. If it is optional, and if your testing shows that it works without it, it is Ok with me because I do not use it.

5.	Could you please also incorporate my changes to ContentMigration.java? These are important and we use these:
a.	A typo in the variable name: worksflowState- > workflowState
b.	migrationIssuesCount

6.	I do not think that I could use your approach with course events. When adding UpdateCourseOptions to publish a course, I did consider using your CourseWriter.updateCourse(Course) method. However, a long time passes between the moment we first create a course and the moment we publish it. We create a course in a separate job, with certain parameters. Then Canvas system administrators may change these parameters through the UI. Then we add course content. And only a while later, when the course is completely ready, we publish it. So I was concerned that creating a new Course object at the time we publish (or even querying it from Canvas) may wipe out some settings or set them back to defaults. This is why I chose to add a CourseWriter.updateCourse(UpdateCourseOptions) method. But I am open to any other suggestions, as long as they do not override the existing course settings and do only one thing: publish a course.

7.	You are right: currently, UpdateCourseOptions can only be used to publish a course. But I was planning to add to it soon to be able to set a course[restrict_enrollments_to_course_dates] option. I would appreciate your direction on how to do it properly.

8.	Are you planning to incorporate my changes to SimpleRestClient.java? These are errors that we successfully retry (with exponential backoff and jitter), but I would appreciate your opinion if you want to handle them differently:
a.	On rare occasions, when we make simultaneous rubrics batch calls, Canvas returns error 403, as described at https://canvas.instructure.com/doc/api/file.throttling.html. Your RateLimitException is thrown when canvas is only “warning” us that it may run out of limit, but catching it does not prevent us from getting 403 on some other calls. 
b.	On some very rare occasions, we also observe error 504 – Gateway timeout. 

9.	I noticed that you merged a pull request for SIS Imports API. Does it only allow to import from a file, using a file path? I already have my csv file loaded into a Java String and would like to use it as an input variable for the call. Is it already possible or do I need to extend your package for that?

Thank you again. Your package is really good!

ToeBee commented:
Answers to your points:

**2**
 Yes, I believe these object mappings are correct. Canvas seems to be needlessly confusing for some of these objects and some of them are not documented at all which is frustrating. And yes, you can still use the `GetSubmissionOptions.RUBRIC_ASSESSMENT` (or `FULL_RUBRIC_ASSESSMENT`) option to get rubric information back when you request `Submission` objects.

**3** 
You are correct. It looks like I just missed that when I did some renaming of things.

**4**
Correct again. The `ignore_for_scoring` field is not listed in the documentation so I missed it. I also see `mastery_points` and `learning_outcome_id` that are not documented but returned in some sample data I have.

**5**
Someone else already made these changes and got them incorporated into the oxctl fork of this project which I have pulled back into this repository. So these changes are already on master.

**6**
By default, the Gson library that we are using to serialize/deserialize objects does not serialize fields that are null. This means that if you create a new blank Course object with only the course ID and the event string, it will only push the event string change to Canvas and will leave the rest of the object  unmodified. Of course, this is a problem if you are explicitly trying to blank something out. So there is an option to change this behavior when creating a new writer object by using the `CanvasApiFactory.getWriter(type, oauthToken, serializeNulls)` method. If you create your writer object with a `true` in the last parameter, then doing this would indeed cause a lot of problems. 

**However**... Maybe changing the course status like this deserves its own method in `CourseWriter` - especially since there is already a `deleteCourse` method. I am still trying to decide if it would be better to make individual methods like `publishCourse(String courseId)`, `unpublishCourse(String courseId)`, etc or maybe a single method like `pushEvent(String courseId, CourseEvent event)` method. The "event" name seems kind of awkward to me. Wouldn't most people call publishing/unpublishing/deleting a course a "status change" or something like that?

**8**
Ah yes, the error handling.
HTTP 403: I was surprised to see someone actually hit the throttling limit! We tried to do it with assignment submissions but were never able to make much of a dent in the rate-limit-remaining header, even when using multiple threads to push several submissions per second. So for our purposes it wasn't a big concern. But I guess assignment submissions are just a cheap call and rubrics are much more expensive. Also, it seemed wrong to me for Canvas to return a 403 because that is supposed to indicate an authorization error. Do you know if it is possible to differentiate between a 403 caused by permission problems and a 403 caused by throttling? Or maybe is Canvas like GitHub in that it just returns a 404 if you don't have permission to view something? I haven't tested with these conditions enough to know. But also, HTTP 429 "Too Many Requests" exists so why didn't they use that?  *sigh* 

HTTP 504: I think we do get this error on occasion as well. I guess I wasn't sure if these requests should be retried or not. When this happens in our own applications (not Canvas) it has usually meant that the load balancer or Apache proxy in front of the application server timed out but the request was still being processed in the application and would finish eventually. So that's why we didn't bother with an exception here. But I think I am ok with including  your change for this

**9**
Hmm. I see the `createSisImport` method takes a `CreateSisImportOptions` object which has both a `filePath` and an `InputStream` parameter. These both get passed down to the `SimpleRestClient` [here](https://github.com/kstateome/canvas-api/blob/master/src/main/java/edu/ksu/canvas/net/SimpleRestClient.java#L179). If the InputStream is null then it creates a new InputStream from the filePath. Otherwise it uses the supplied InputStream. So it *seems* like you could create an `InputStream` from your string in memory and pass it in there. I'm not sure if you would still need a `filePath` parameter or not. I *think* this is just passed along as an informational thing but I have not tested this code myself yet.

It is bed time here but I will try to work some more on this PR tomorrow. Thank you for your input!

ToeBee commented:
@alina-spectorov ok, could you please look this over again? Changes:

**Class fields**
I fixed my renaming error in `Rubric` and updated the `RubricCriterion` class with more fields.

**Error handling**
 I have added your error handling changes back in. My concern about a 403 possibly having multiple meanings was invalid. It turns out, if you have a valid token but it does not have permission to view a requested resource, Canvas returns a 401, not a 403. Of course this overloads the 401 response for both invalid token problems and lack of permission. However this is differentiated with the `WWW-Authenticate` header which the SimpleRestClient class already handles. I did change the name of your exception class to just be `ThrottlingException` instead of `CanvasThrottlingException` but other than that, it is the same as your code.

**Course update**
While I was debating how to do this, I noticed a different API call and decided to implement that instead. Depending on how you are publishing your courses, this might be even better than anything we had talked about previously... Or it might be annoying.

There is a batch update for courses ([see here](https://canvas.instructure.com/doc/api/courses.html#method.courses.batch_update)) that only lets you update the workflow state. Since you can update multiple courses at once, it doesn't return a Course object but rather a `Progress` object that lets you monitor the progress of the batch job. So I implemented the following method to call this API endpoint:  
`CourseWriter.batchUpdateCourseState(String accountId, Course.CourseEvent event, String... courseIds)`

You can pass in a single course or a whole list of dozens of courses to publish/conclude/etc at once. But if you are only doing one course at a time, then getting that Progress object back might be annoying.

The problem with your approach of putting the event (and then `restrict_enrollments_to_course_dates`) into an `UpdateCourseOptions` object is that then we will eventually be duplicating all the fields in the `Course` object needlessly. I do understand that it is a little weird to have to create a blank Course object with nothing but an ID and one field in it and push that to Canvas. I think in my mind I was thinking that people would *get* the course from Canvas, change things as needed and then push the whole thing back. However that is definitely not always a great workflow.

Thoughts?

alina-spectorov commented:
Hi Toby,

1.	I am glad that my comments about Rubric class fields helped. I received emails about your commits and saw that you incorporated them.

2.	Thank you for adding error handling and for your thoughtful research on it.

3.	Good catch with courses batch update call! I think there is value in adding it to your package regardless of whether I end up using it. For our particular workflow, I still have concerns:

  a.	I may have to pass in a single course because we publish each course in a separate thread, after checking on its content migration workflow state and issues count.

  b.	Currently, after we publish a course, we immediately check on its workflow state. In a highly unlikely event that the course is not published, we can add it to the list and notify our system admins. If I switch to using the batch update, does it mean that Canvas will queue it for later (like it does with sis imports and content migrations)? If yes, I would need to create and to maintain a separate automated job to check on published courses later.

  c.	In the future, if I need to update a course[restrict_enrollments_to_course_dates] option (this project is not approved yet), do you recommend to create a blank Course object with nothing but an ID and one field in it and to push that to Canvas? Maybe in this case I would need to call CanvasApiFactory.getWriter(type, oauthToken, serializeNulls) with a false in the last parameter set explicitly? What do you think? 
            - Is there a guarantee that the writer object will always honor the request not to serialize nulls? 
            - Does pushing the whole object to Canvas do a POST or a PUT?  I think we need a PUT for an update request. From following your code, I think that we are Ok here, but could you please confirm?

Thank you, Alina.

ToeBee commented:
Yeah, I was afraid you might have a workflow like that so the batch update may not be what you want. You would have to push a batch and then check the state of all the courses after the batch job has completed.

Not serializing null fields is the default behavior for the the Gson library. After I first released this library there was no way to UNSET things in Canvas. So I had to add the option to force Gson to serialize nulls to allow for setting fields to blank. I do not think the default behavior will change but you can certainly pass in `false` to the `getWriter` method to make this behavior explicit.

You are correct that update methods send a PUT request. Create methods send a POST.

So I think I am going to merge this and prepare for a release!

alina-spectorov commented:
Thank you!

dmalia1 commented on src/main/java/edu/ksu/canvas/impl/CourseImpl.java:
```diff
@@ -40,14 +43,14 @@ public CourseImpl(String canvasBaseUrl, Integer apiVersion, OauthToken oauthToke
 
     public List<Course> listUserCourses(ListUserCoursesOptions options) throws  IOException {
-        LOG.debug("listing course for user");
+        LOG.debug("listing course for user {}", options.getUserId());
```
If options is null, this could throw a NullPointerException.

ToeBee commented:
While it is technically possible for the options objects to be null, in practice I don't think this is a big concern because of how they are used. You typically construct an options object immediately before using it and it is created for a specific API call. So I feel like it would be somewhat difficult to accidentally pass in a null. I would not be opposed to adding NotNull annotations but if I do that, it should be to all of the API calls at once so that things are consistent. I will put it on a list of refactorings to do for 2.0.
